core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1664,11 +1664,21 @@ class KafkaController(val config: KafkaConfig,
   }
 
   private def processTopicIds(topicIdAssignments: Set[TopicIdReplicaAssignment]): Unit = {
-    // Create topic IDs for topics missing them if we are using topic IDs
+    // Create topic IDs or update with locally stored topicIDs for topics missing them if we are using topic IDs
     // Otherwise, maintain what we have in the topicZNode
     val updatedTopicIdAssignments = if (config.usesTopicId) {
-      val (withTopicIds, withoutTopicIds) = topicIdAssignments.partition(_.topicId.isDefined)
-      withTopicIds ++ zkClient.setTopicIds(withoutTopicIds, controllerContext.epochZkVersion)
+      val (withTopicIds, withoutTopicIds, withLocalTopicIds) = topicIdAssignments.foldLeft((Set.empty[TopicIdReplicaAssignment], Set.empty[TopicIdReplicaAssignment], Set.empty[TopicIdReplicaAssignment])) {
```

Comment:
readability is a bit tricky here. I'm also wondering if there was an operation you could have done besides making singleton sets and unioning them.

Reply:
Since we go through all the partitions again to check the topic ID and add to the replica assignment I wonder if we could have kept the partition and then a separate part for the ones with local IDs. 🤔

Reply:
> I'm also wondering if there was an operation you could have done besides making singleton sets and unioning them.

I chose this approach because this was the only one where we are able to 3-way partition the `topicIdAssignments` in a single iterations. All other approaches require multiple iterations.

We can work on improving this and favouring readibility once we have a consensus on whether we will proceed with this change or not.